propagate_const.h

```diff
@@ -41,84 +67,107 @@ class propagate_const {
+  PROPAGATE_CONST_CONSTEXPR propagate_const() = default;
 
   propagate_const(const propagate_const& p) = delete;
 
-  constexpr propagate_const(propagate_const&& p) = default;
+  PROPAGATE_CONST_CONSTEXPR propagate_const(propagate_const&& p) = default;
 
+#ifdef PROPAGATE_CONST_HAS_EXPRESSION_SFINAE
+  //
+  // Use SFINAE to check if converting constructor should be explicit.
+  //
   template <class U, enable_if_t<!is_convertible<U&&, T>::value &&
                                      is_constructible<T, U&&>::value,
                                  bool> = true>
-  explicit constexpr propagate_const(propagate_const<U>&& pu)
+  explicit PROPAGATE_CONST_CONSTEXPR propagate_const(propagate_const<U>&& pu)
       : t_(std::move(pu.t_)) {}
 
   template <class U, enable_if_t<is_convertible<U&&, T>::value &&
                                      is_constructible<T, U&&>::value,
                                  bool> = false>
-  constexpr propagate_const(propagate_const<U>&& pu) : t_(std::move(pu.t_)) {}
+  PROPAGATE_CONST_CONSTEXPR propagate_const(propagate_const<U>&& pu) : t_(std::move(pu.t_)) {}
 
   template <class U, enable_if_t<!is_convertible<U&&, T>::value &&
                                      is_constructible<T, U&&>::value &&
                                      !is_propagate_const<decay_t<U>>::value,
                                  bool> = true>
-  explicit constexpr propagate_const(U&& u) : t_(std::forward<U>(u)) {}
+  explicit PROPAGATE_CONST_CONSTEXPR propagate_const(U&& u) : t_(std::forward<U>(u)) {}
 
   template <class U, enable_if_t<is_convertible<U&&, T>::value &&
                                      is_constructible<T, U&&>::value &&
                                      !is_propagate_const<decay_t<U>>::value,
                                  bool> = false>
-  constexpr propagate_const(U&& u) : t_(std::forward<U>(u)) {}
+  PROPAGATE_CONST_CONSTEXPR propagate_const(U&& u) : t_(std::forward<U>(u)) {}
+
+#else
+  //
+  // Make converting constructors explicit as we cannot use SFINAE to check.
+  //
+  template <class U, class = enable_if_t<is_constructible<T, U&&>::value>>
+  explicit PROPAGATE_CONST_CONSTEXPR propagate_const(propagate_const<U>&& pu)
+      : t_(std::move(pu.t_))
+  {
+  }
 
+  template <class U,
+            class = enable_if_t<is_constructible<T, U&&>::value &&
+                                !is_propagate_const<decay_t<U>>::value>>
+  explicit PROPAGATE_CONST_CONSTEXPR propagate_const(U&& u)
+      : t_(std::forward<U>(u))
+  {
+  }
+
+#endif
   // [propagate_const.assignment], assignment
   propagate_const& operator=(const propagate_const& p) = delete;
 
-  constexpr propagate_const& operator=(propagate_const&& p) = default;
+  PROPAGATE_CONST_CONSTEXPR propagate_const& operator=(propagate_const&& p) = default;
 
   template <class U>
-  constexpr propagate_const& operator=(propagate_const<U>&& pu) {
+  PROPAGATE_CONST_CONSTEXPR propagate_const& operator=(propagate_const<U>&& pu) {
```

Comment:
Wondering if this overload shouldn't be constrained when the equivalent construction is implicit.

Reply:
I think you are right. 

I'll improve the tests next. Then we can iterate on improving things.
Tests from libcxx are written by me and easy to re-purpose.